### PR TITLE
First draft of document comparing upb's design with C++ protos.

### PR DESCRIPTION
**Pull request by haberman**
Staged at: https://github.com/haberman/upb/blob/vs-doc/doc/vs-cpp-protos.md

### CODE REVIEW

File: doc/vs-cpp-protos.md [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+  data races in a thread-compatible arena like upb.
+* A thread-compatible arena has a far simpler implementation.  The C++ thread-safe
+  arena relies on thread-local variables, which introduce complications on some
+  platforms.  It also requires far more subtle reasoning for correctness and
+  performance, and likely cannot match the performance of a thread-compatible arena.
+
+**fuse vs. no fuse**
+* The `upb_Arena_Fuse()` operation is a key part of how we can support reparenting
+  of messages when the parent may be on a different arena.  Without this, we have
+  no way of supporting `foo.bar = bar` in dynamic languages without performing a
```
Comment (fowles) [inline doc/vs-cpp-protos.md:118]: s/we have/upb has/

File: doc/vs-cpp-protos.md [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+
+### upb
+
+upb does not generate any code into `foo.upb.c` files, only data structures.  upb uses a
+compact data table known as a *mini table* to represent the schema and all fields.
+
+upb uses mini tables to perform all of the operations that would traditionally be done
+with generated code.  Revisiting the list from the previous section:
+
+* `FooMsg::FooMsg()` (constructor): upb instead initializes all messages with `memset(msg, 0, size)`.
```
Comment (fowles) [inline doc/vs-cpp-protos.md:157]: random side thought: it seems like you should be able to memset only the hasbits

File: doc/vs-cpp-protos.md [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+This means we are seeing both the overhead of the core library itself as well as the
+generated code (or table) for `descriptor.proto`.  (For extra clarity we should break this
+down by generated code vs core library in the future).
+
+
+| Library         | `.text` | `.data` | `.bss` |
+|------------     |---------|---------|--------|
+| upb             |  26Ki   | 0.6Ki   | 0.01Ki |
+| C++ (lite)      | 187Ki   | 2.8Ki   | 1.25Ki |
+| C++ (code size) | 904Ki   | 6.1Ki   | 1.88Ki |
```
Comment (fowles) [inline doc/vs-cpp-protos.md:181]: I assume by (code size) you mean compiling with `-Os` on the full runtime, but you should say that

Comment (haberman) [reply doc/vs-cpp-protos.md:181]: No that is `optimize_for = CODE_SIZE`, a mode in which `.pb.cc` contains descriptors only (no generated code). I'll add a note. I think I want to rewrite this section anyway, will send something soon.

**Review by fowles — APPROVED** [review]

**Pull request merged**